src/hangar/backends/hdf5.py

```diff
@@ -574,20 +574,36 @@ def read_data(self, hashVal: HDF5_00_Parser.DataHashSpec) -> np.ndarray:
 
         srcSlc = (self.slcExpr[dsetIdx], *(self.slcExpr[0:x] for x in hashVal.shape))
         destSlc = None
-        destArr = np.empty((hashVal.shape), self.schema_dtype)
 
-        try:
-            self.Fp[hashVal.uid][dsetCol].read_direct(destArr, srcSlc, destSlc)
-        except TypeError:
-            self.Fp[hashVal.uid] = self.Fp[hashVal.uid]()
-            self.Fp[hashVal.uid][dsetCol].read_direct(destArr, srcSlc, destSlc)
-        except KeyError:
-            file_pth = pjoin(self.STAGEDIR, f'{hashVal.uid}.hdf5')
-            if (self.mode == 'a') and os.path.islink(file_pth):
-                self.rFp[hashVal.uid] = h5py.File(file_pth, 'r', swmr=True, libver='latest')
+        if self.schema_dtype is not None:
+            destArr = np.empty((hashVal.shape), self.schema_dtype)
+            try:
+                self.Fp[hashVal.uid][dsetCol].read_direct(destArr, srcSlc, destSlc)
+            except TypeError:
+                self.Fp[hashVal.uid] = self.Fp[hashVal.uid]()
                 self.Fp[hashVal.uid][dsetCol].read_direct(destArr, srcSlc, destSlc)
-            else:
-                raise
+            except KeyError:
+                process_dir = self.STAGEDIR if self.mode == 'a' else self.STOREDIR
+                file_pth = pjoin(process_dir, f'{hashVal.uid}.hdf5')
+                if os.path.islink(file_pth):
+                    self.rFp[hashVal.uid] = h5py.File(file_pth, 'r', swmr=True, libver='latest')
+                    self.Fp[hashVal.uid][dsetCol].read_direct(destArr, srcSlc, destSlc)
+                else:
+                    raise
+        else:
+            try:
+                destArr = self.Fp[hashVal.uid][dsetCol][srcSlc]
```

Comment:
This is an optimization for HDF5. 

In the first case the schema hash is known; this lets us know the `dtype` and `shape` of the before we look at the data on disk at all. This is beneficial because we can an empty array with the appropriate parameters at the start. By using the `read_direct` function, we allow the HDF5 library to avoid making intermediate copies of the data as each `chunk` is read from disk (after which the each `chunk` would normally be stitched together and a contiguous array returned to the user). With `read_direct`, each `chunk` is moved straight from the disk into its appropriate location in the `destination` array. It's a bit tedious as you see here, but it results in roughly a 30% - 40% speedup.

If the data exists on the local coputer, we almost always know the schema. However, there are situations in the remote `client` / `server` operations where the processing time required to determine the schema of some data digest exceeds the speedup provided by the `read_direct` operation. 

does this answer your question? 

F

Reply:
Nice, thanks